app/src/hid/hid_gamepad.c

```diff
@@ -0,0 +1,456 @@
+#include "hid_gamepad.h"
+
+#include <assert.h>
+#include <inttypes.h>
+
+#include "util/binary.h"
+#include "util/log.h"
+
+// 2 bytes for left stick (X, Y)
+// 2 bytes for right stick (Z, Rz)
+// 2 bytes for L2/R2 triggers
+// 2 bytes for buttons + padding,
+// 1 byte for hat switch (dpad) + padding
+#define SC_HID_GAMEPAD_EVENT_SIZE 15
+
+// The ->buttons field stores the state for all buttons, but only some of them
+// (the 16 LSB) must be transmitted "as is". The DPAD (hat switch) are stored
+// locally in the MSB, but not transmitted as is: they are transformed to
+// generate another specific byte.
+#define SC_HID_BUTTONS_MASK 0xFFFF
+
+// outside SC_HID_BUTTONS_MASK
+#define SC_GAMEPAD_BUTTONS_BIT_DPAD_UP    UINT32_C(0x10000)
+#define SC_GAMEPAD_BUTTONS_BIT_DPAD_DOWN  UINT32_C(0x20000)
+#define SC_GAMEPAD_BUTTONS_BIT_DPAD_LEFT  UINT32_C(0x40000)
+#define SC_GAMEPAD_BUTTONS_BIT_DPAD_RIGHT UINT32_C(0x80000)
+
+/**
+ * Gamepad descriptor manually crafted to transmit the events.
+ *
+ * The HID specification is available here:
+ * <https://www.usb.org/document-library/device-class-definition-hid-111>
+ *
+ * The HID Usage Tables is also useful:
+ * <https://www.usb.org/document-library/hid-usage-tables-15>
+ */
+static const uint8_t SC_HID_GAMEPAD_REPORT_DESC[] = {
+    // Usage Page (Generic Desktop)
+    0x05, 0x01,
+    // Usage (Gamepad)
+    0x09, 0x05,
+
+    // Collection (Application)
+    0xA1, 0x01,
+
+    // Collection (Physical)
+    0xA1, 0x00,
+
+    // Usage Page (Generic Desktop)
+    0x05, 0x01,
+    // Usage (X)   Left stick x
+    0x09, 0x30,
+    // Usage (Y)   Left stick y
+    0x09, 0x31,
+    // Usage (Z)   Right stick x
+    0x09, 0x32,
+    // Usage (Rz)  Right stick y
+    0x09, 0x35,
+    // Logical Minimum (0)
+    0x15, 0x00,
+    // Logical Maximum (65535)
+    // Cannot use 26 FF FF because 0xFFFF is interpreted as signed 16-bit
+    0x27, 0xFF, 0xFF, 0x00, 0x00, // little-endian
+    // Report Size (16)
+    0x75, 0x10,
+    // Report Count (4)
+    0x95, 0x04,
+    // Input (Data, Variable, Absolute): 4 bytes (X, Y, Z, Rz)
+    0x81, 0x02,
+
+    // Usage Page (Simulation Controls)
+    0x05, 0x02,
+    // Usage (Brake)
+    0x09, 0xC5,
+    // Usage (Accelerator)
+    0x09, 0xC4,
+    // Logical Minimum (0)
+    0x15, 0x00,
+    // Logical Maximum (65535)
+    0x27, 0xFF, 0xFF, 0x00, 0x00,
+    // Report Size (16)
+    0x75, 0x10,
+    // Report Count (2)
+    0x95, 0x02,
+    // Input (Data, Variable, Absolute): 2 bytes (L2, R2)
+    0x81, 0x02,
+
+    // Usage Page (Buttons)
+    0x05, 0x09,
+    // Usage Minimum (1)
+    0x19, 0x01,
+    // Usage Maximum (15)
+    0x29, 0x0E,
```

Comment:
Right stick button doesn't work. This should be `0x29, 0x0F`

Reply:
Thank you :+1:

Reply:
In fact, what was important was:
```diff
-    // Report Count (15)
-    0x95, 0x0F,
+    // Report Count (16)
+    0x95, 0x10,
```

Here, the value should even be `0x29, 0x10` (not `0x0F`). But it seems harmless. If fix it locally.